pkg/upgrade/upgrade.go

```diff
@@ -0,0 +1,246 @@
+}
+
+func GetBinPath(pid int) (string, error) {
+	_, err := os.FindProcess(pid)
+	if err != nil {
+		return "", fmt.Errorf(" process is not found")
+	}
+	if pid == 1 {
+		return "/bin/k3s", nil
+	}
+	binPath, err := ioutil.ReadFile(path.Join("/host/proc", strconv.Itoa(pid), "cmdline"))
+	if fileExists(string(binPath)) {
+		return string(binPath), nil
+	}
+	return "", fmt.Errorf(" failed to fetch the k3s binary path from process %d", pid)
+}
+
+func TermProcess(pid int) error {
+	proc, err := os.FindProcess(pid)
+	if err != nil {
+		fmt.Errorf(" process is not found")
+	}
+	return proc.Signal(syscall.SIGTERM)
+}
+
+func MatchSelector(hostPath, configPath string, preservedEntries []string) error {
+	kubeconfig := os.Getenv("KUBECONFIG")
+	conf, err := clientcmd.BuildConfigFromFlags("", kubeconfig)
+	if err != nil {
+		return err
+	}
+	nodeClient := kubernetes.NewForConfigOrDie(conf).CoreV1().Nodes()
+	hostname := os.Getenv("SYSTEM_UPGRADE_NODE_NAME")
+	if hostname == "" {
+		hostname, err = os.Hostname()
+		if err != nil {
+			return err
+		}
+	}
+	host, err := nodeClient.Get(context.Background(), hostname, v1.GetOptions{})
+	if err != nil {
+		return err
+	}
+	hostLabels := host.GetLabels()
+	secretfiles, err := ioutil.ReadDir(configPath)
+	if err != nil {
+		return err
+	}
+	var secret types.Secret
+	secretfiles = sortFileByName(secretfiles)
+	for _, secretFile := range secretfiles {
+		fileContent, err := ioutil.ReadFile(secretFile.Name())
+		if err != nil {
+			return err
+		}
+		yaml.Unmarshal(fileContent, secret)
+		selector, err := v1.LabelSelectorAsSelector(&secret.Selector)
+		if err != nil {
+			return err
+		}
+		if selector.Matches(labels.Set(hostLabels)) {
+			//check if a config file is present on the node
+			return PlaceConfigFile(hostPath, secret.Config, preservedEntries)
+
+		}
+	}
+	return nil
+}
+
+func PlaceConfigFile(hostPath string, config map[string]interface{}, preservedEntries []string) error {
+	preserved := make(map[string]interface{})
+	if fileExists(hostPath) {
+		logrus.Infof("config file exists on the host")
+		filetxt, err := ioutil.ReadFile(hostPath)
+		if err != nil {
+			return err
+		}
+		hostconfig := make(map[string]interface{})
+		err = yaml.Unmarshal(filetxt, hostconfig)
+		if err != nil {
+			return err
+		}
+		for _, entry := range preservedEntries {
+			if hostconfig[entry] != nil {
+				preserved[entry] = hostconfig[entry]
+			}
+		}
+	}
+
+	for entry := range preserved {
+		config[entry] = preserved[entry]
+	}
+	content, err := yaml.Marshal(config)
+	if err != nil {
+		return err
+	}
+	return ioutil.WriteFile(hostPath, content, 0755)
+}
+
+func sortFileByName(files []os.FileInfo) []os.FileInfo {
+	sort.Slice(files, func(i, j int) bool {
+		return files[i].Name() < files[j].Name()
+	})
+	return files
+}
+
+func ReplaceBinary() error {
+	newBinaryPath := "/opt/k3s"
+	pid, err := GetProcessPID()
+	if err != nil {
+		return err
+	}
+	fullBinPath, err := GetBinPath(pid)
```

Comment:
I am worried that this is returning a string with null bytes in it. See my above comment on `GetBinPath()`